tests/Compiler/Type/PhpDocTypeUtilsTest.php

```diff
@@ -417,4 +422,404 @@ public function testResolve(): void
         self::assertSame(self::class, $identifier->name);
     }
 
+    /**
+     * @param list<string> $true
+     * @param list<string> $false
+     */
+    #[DataProvider('provideIsSubTypeOfData')]
+    public function testIsSubTypeOf(string $type, array $true, array $false): void
+    {
+        $typeNode = $this->parseType($type);
+
+        foreach ($true as $trueType) {
+            $trueTypeNode = $this->parseType($trueType);
+            self::assertTrue(PhpDocTypeUtils::isSubTypeOf($trueTypeNode, $typeNode), "$trueType is not subtype of $type");
+        }
+
+        foreach ($false as $falseType) {
+            $falseTypeNode = $this->parseType($falseType);
+            self::assertFalse(PhpDocTypeUtils::isSubTypeOf($falseTypeNode, $typeNode), "$falseType is subtype of $type");
+        }
+    }
+
+    private function parseType(string $type): TypeNode
+    {
+        $lexer = new Lexer();
+        $constExprParser = new ConstExprParser(unescapeStrings: true);
+        $typeParser = new TypeParser($constExprParser);
+
+        return $typeParser->parse(new TokenIterator($lexer->tokenize($type)));
+    }
+
+    /**
+     * @return iterable<string, array{type: string, true: list<string>, false: list<string>}>
+     */
+    public static function provideIsSubTypeOfData(): iterable
+    {
+        yield 'array' => [
+            'type' => 'array',
+
+            'true' => [
+                'array',
+                'int[]',
+                'array{int}',
+                'list<int>',
+                'array<int>',
+                'array<int, int>',
+            ],
+
+            'false' => [
+                'int',
+                'iterable',
+            ],
+        ];
+
+        yield 'array<int>' => [
+            'type' => 'array<int>',
+
+            'true' => [
+                'array<int>',
+                'array<int, int>',
+                'list<int>',
+                'array{int}',
+                'int[]',
+            ],
+
+            'false' => [
+                'array',
+                'array{int, string}',
+                'int',
+            ],
+        ];
+
+        yield 'array<string, int>' => [
+            'type' => 'array<string, int>',
+
+            'true' => [
+                'array<string, int>',
+                'array{foo: int}',
+                'array{"foo": int}',
+            ],
+
+            'false' => [
+                'array',
+                'array<int, string>',
+                'array{string}',
+                'array{123: string}',
+                'int',
+            ],
```

Comment:
What about unions inside? Like `array{'foo'|'bar'}`, `array{1|2: string}`. Not sure if that is supported by parser though.

Reply:
That is not supported by the parser. Key must be one of `ConstExprIntegerNode|ConstExprStringNode|IdentifierTypeNode|null`.